### PR TITLE
jira: general client improvements

### PR DESCRIPTION
**Pull request by AlexNPavel**
This PR fixes the fakejira `CloneIssue` and `CreateIssue` functions
along with adding unit tests for them, adds more extensive issue
function errors for the fakeclient, and modifies the
`DeleteRemoteLinkViaURL` function to return a `bool` indicating whether
a link with the provided URL was found and removed (no remote link with
the provided URL is not an error).

These improvements will be used in a new openshift external prow plugin
called the `jira-lifecycle` plugin, which will replicate the `bugzilla`
plugins functionality, but with the Jira service instead of bugzilla.

/cc @stevekuznetsov @alvaroaleman

### CODE REVIEW

File: prow/jira/fakejira/fake.go [hunk 1]
```diff
@@ -116,44 +131,73 @@ func (f *FakeClient) CreateIssueLink(link *jira.IssueLink) error {
 	if err != nil {
 		return fmt.Errorf("failed to get outward link issue: %v", err)
 	}
-	outward.Fields.IssueLinks = append(outward.Fields.IssueLinks, link)
+	// links in an issue struct do not include the short definition of the the issue they are part of.
+	// This behavior is used by the external jira-lifecycle hook plugin to identify which direction a clone happened,
```
Comment (stevekuznetsov) [inline prow/jira/fakejira/fake.go:135]: I also don't think comments about downstream consumers of the library are a good idea here. Regardless of who uses this, the link is a valid thing to set, right?

File: prow/jira/fakejira/fake.go [hunk 2]
```diff
@@ -116,44 +131,73 @@ func (f *FakeClient) CreateIssueLink(link *jira.IssueLink) error {
...
-		return nil, err
-	}
-	// set ID and Key to unused id and key
-	f.updateIssueIDAndKey(issueCopy)
-	// run generic cloning function
-	return jiraclient.CloneIssue(f, issueCopy)
+	return jiraclient.CloneIssue(f, issue)
 }
 
 func (f *FakeClient) CreateIssue(issue *jira.Issue) (*jira.Issue, error) {
```
Comment (stevekuznetsov) [inline prow/jira/fakejira/fake.go:155]: What, exactly, is this helping? Having to build a fake Jira sounds like testing is happening at the wrong level. See also: how this repo has not built a fake GitHub, Bugzilla, S3, etc.

File: prow/jira/fakejira/fake_test.go [hunk 3]
```diff
@@ -72,3 +78,203 @@ func TestFakeClient_SearchWithContext(t *testing.T) {
...
+			},
+		},
+	}}
+	for _, tc := range testCases {
+		ptrIssues := []*jira.Issue{}
+		for index := range tc.existingIssues {
+			ptrIssues = append(ptrIssues, &tc.existingIssues[index])
+		}
+		fakeClient := &FakeClient{Issues: ptrIssues}
+		newIssue, err := fakeClient.CreateIssue(&tc.issue)
```
Comment (stevekuznetsov) [inline prow/jira/fakejira/fake_test.go:173]: These are unit tests for a fake ... ? This doesn't seem right - in the fake you might inject data, or validate that calls happen to the right endpoints. There's nothing making sure that the fake implementation of the clone here is the same as what Jira actually does, so ensuring it is "correct" here is not helpful.

File: prow/jira/fakejira/fake.go (hunk 2 above)
Comment (AlexNPavel) [reply prow/jira/fakejira/fake.go:154]: What do you mean by this repo not building a fake github or bugzilla? There is a fakegithub and a fake bugzilla client that are used heavily by unit testing in plugins. This fake jira client isn't much different.

Comment (stevekuznetsov) [reply prow/jira/fakejira/fake.go:154]: Those are simple in-memory maps that we add to / delete from using the methods. They do not encode provider-specific implementation details like `("issue.fields.project must be set to create new issue")`.

File: prow/jira/fakejira/fake_test.go (hunk 3 above)
Comment (AlexNPavel) [reply prow/jira/fakejira/fake_test.go]: I have removed the unit tests.

**Pull request merged**